File: 39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> res;
    vector<int> cur;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        backtrack(candidates, target, 0, 0);
        return res;
    }

    void backtrack(vector<int>& candidates, int targetSum, int sum, int index) {
        if(sum > targetSum) return;
        if(sum == targetSum) {
            res.push_back(cur);
            return;
        }

        for(int i = index; i < candidates.size() && candidates[i] + sum <= targetSum; i++) {
            cur.push_back(candidates[i]);
            backtrack(candidates, targetSum, sum + candidates[i], i);
            cur.pop_back();
        }

    }

};
// @lc code=end

File: 39.组合总和.py
#
# @lc app=leetcode.cn id=39 lang=python3
#
# [39] 组合总和
#

# @lc code=start
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.res = []
        path = []
        def backtrack(index, current_sum):
            if current_sum == target:
                self.res.append(path.copy())
                return
            
            for i in range(index, len(candidates)):
                current_sum += candidates[i]
                if current_sum > target: 
                    current_sum -= candidates[i]
                    continue
                path.append(candidates[i])
                backtrack(i, current_sum)
                current_sum -= candidates[i]
                path.pop()
        backtrack(0, 0)
        return self.res
        


            
                    
                
        
# @lc code=end

